File: packages/quran-engine-kotlin/src/main/kotlin/com/quranengine/Names.kt
package com.quranengine

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * The 99 Names of Allah (Asma' ul-Husna). Ported from `packages/quran-engine-js/src/names.js`.
 *
 * Thin accessor over `data/names-of-allah.json`.
 */

/** A single Name of Allah. Mirrors the JS `NameOfAllah` typedef and `names-of-allah.json[]`. */
@Serializable
data class NameOfAllah(
    @SerialName("name") val name: String,                       // Arabic
    @SerialName("transliteration") val transliteration: String,
    @SerialName("number") val number: Int,                      // 1..99
    @SerialName("found") val found: String = "",                // ayah references, e.g. "(1:3) (17:110)"
    @SerialName("meaning") val meaning: String = "",
    @SerialName("desc") val desc: String = "",
    @SerialName("otherNames") val otherNames: List<String> = emptyList(),
)

/** Accessor over the 99 Names, ordered by `number`. */
class NamesOfAllah(list: List<NameOfAllah> = emptyList()) {
    /** All names, ordered by number. */
    private val list: List<NameOfAllah> = list.sortedBy { it.number }

    private val byNumber: Map<Int, NameOfAllah> = this.list.associateBy { it.number }

    /** All 99 names, ordered by number. */
    fun all(): List<NameOfAllah> = list

    /** The name with the given [number] (1..99), or null if unknown. */
    fun byNumber(number: Int): NameOfAllah? = byNumber[number]
}

File: packages/quran-engine-py/quran_engine/names.py
"""The 99 Names of Allah (Asma' ul-Husna). Thin accessor over data/names-of-allah.json.
Mirrors src/names.js."""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class NameOfAllah:
    name: str               # Arabic
    transliteration: str
    number: int             # 1..99
    found: str = ""         # ayah references where it appears (e.g. "(1:3) (17:110)")
    meaning: str = ""
    desc: str = ""
    other_names: list[str] = field(default_factory=list)

    @staticmethod
    def from_json(d: dict) -> "NameOfAllah":
        return NameOfAllah(
            name=d.get("name", ""),
            transliteration=d.get("transliteration", ""),
            number=d["number"],
            found=d.get("found", ""),
            meaning=d.get("meaning", ""),
            desc=d.get("desc", ""),
            other_names=d.get("otherNames", []),
        )


class NamesOfAllah:
    def __init__(self, names: Optional[list[NameOfAllah]] = None):
        self.list = sorted(names or [], key=lambda n: n.number)
        self._by_number = {n.number: n for n in self.list}

    def all(self) -> list[NameOfAllah]:
        """All 99 names, ordered by number."""
        return self.list

    def by_number(self, number: int) -> Optional[NameOfAllah]:
        return self._by_number.get(number)
